pkg/codegen/schema.go

```diff
@@ -541,29 +541,83 @@ func MergeSchemas(allOf []*openapi3.SchemaRef, path []string) (Schema, error) {
 	return outSchema, nil
 }
 
+// GetReferencedSchema requires the package scope variable Schemas to have been populated
+func GetReferencedSchema(name string) (*openapi3.SchemaRef, error) {
+	if s, ok := Schemas[name]; ok {
+		if s.Value != nil {
+			return s, nil
+		}
+		return nil, fmt.Errorf("schema %s had no value", name)
+	}
+	return nil, fmt.Errorf("could not find referenced schema %s", name)
+}
+
 // This function generates an object that is the union of the objects in the
 // input array. In the case of Ref objects, we use an embedded struct, otherwise,
 // we inline the fields.
 func GenStructFromAllOf(allOf []*openapi3.SchemaRef, path []string) (string, error) {
+	// Check how many refs and props we have
+	refCount := 0
+	savedRef := ""
+	propCount := 0
+	for _, schemaOrRef := range allOf {
+		ref := schemaOrRef.Ref
+		if IsGoTypeReference(ref) {
+			refCount++
+			savedRef = ref
+		} else {
+			goSchema, err := GenerateGoSchema(schemaOrRef, path)
+			if err != nil {
+				return "", err
+			}
+			propCount += len(GenFieldsFromProperties(goSchema.Properties))
+		}
+	}
+	if refCount == 1 && propCount == 0 {
+		// We have just one reference and no properties
+		// so flatten it and just use the referenced struct.
+		goType, err := RefPathToGoType(savedRef)
+		if err != nil {
+			return "", err
+		}
+		return goType, nil
+	}
+
 	// Start out with struct {
 	objectParts := []string{"struct {"}
+
 	for _, schemaOrRef := range allOf {
 		ref := schemaOrRef.Ref
 		if IsGoTypeReference(ref) {
-			// We have a referenced type, we will generate an inlined struct
-			// member.
-			// struct {
-			//   InlinedMember
-			//   ...
-			// }
 			goType, err := RefPathToGoType(ref)
 			if err != nil {
 				return "", err
 			}
-			objectParts = append(objectParts,
-				fmt.Sprintf("   // Embedded struct due to allOf(%s)", ref))
-			objectParts = append(objectParts,
-				fmt.Sprintf("   %s `yaml:\",inline\"`", goType))
+			if true {
```

Comment:
Obviously this needs to switch based on something, perhaps an extension e.g. `x-oapi-codegen-flatten-allof: true`